Physical server provisioning via StateMachine

miha-plesko opened the pull request:
Hi guys, I'm sending in a series of PRs that all aim to start provisioning physical servers by means of a new StateMachine which is supposed to also take care of different kind of physical servers (talking Lenovo XClarity vs RedFish here). It's still a WIP, but the code alread works. 

- [ ] https://github.com/ManageIQ/manageiq-providers-redfish/pull/58 (React Component)
- [x] https://github.com/ManageIQ/manageiq-providers-redfish/pull/60 (with_provider_object)
- [x] https://github.com/ManageIQ/manageiq-api/pull/578
- [x] https://github.com/ManageIQ/manageiq/pull/18573
- [x] https://github.com/ManageIQ/manageiq-automation_engine/pull/306
- [x] https://github.com/ManageIQ/manageiq-content/pull/516
- [x] https://github.com/ManageIQ/manageiq-ui-classic/pull/5461

---
On this PR we implement PhysicalServer provisioning (internal) state machine. Provisioning is done in three steps:

```
- deploy_pxe_config
- reboot_using_pxe
- poll_server_running
```

We assume provisioning request is triggered with following options:

```
POST /api/requests
{
  "options" : {
	"request_type" : "provision_physical_server",
	"src_ids": [1, 2, 3],           # list of PhysicalServers to provision
	"pxe_image_id": 1,              # what PXE Image to use
	"customization_template_id": 1  # what customization template to use
  }
}
```

Please note that the internal state machine (implemented in this commit) is executed per src_id. In the example above, the internal state machine would get executed three times: once for PhysicalServer id=1, once for id=2 and once for id=3.

Please also note that each phase of the state machine is lightweight - because a single worker can only grab&execute one phase at a time which means a long-running phase blocks all other state machines.

gmcculloug commented on app/models/manageiq/providers/redfish/physical_infra_manager/physical_server/operations.rb:
```diff
@@ -0,0 +1,62 @@
+  module PhysicalInfraManager::PhysicalServer::Operations
+    def provision(pxe_image_id, customization_template_id)
+      image = PxeImage.find(pxe_image_id)
+      template = CustomizationTemplate.find(customization_template_id)
```
Use `find_by(:id =>`

```
image = PxeImage.find_by(:id => pxe_image_id)
template = CustomizationTemplate.find_by(:id => customization_template_id)
```

gmcculloug commented on app/models/manageiq/providers/redfish/physical_infra_manager/physical_server/operations.rb:
```diff
@@ -0,0 +1,62 @@
+require "set"
+
+module ManageIQ::Providers::Redfish
+  module PhysicalInfraManager::PhysicalServer::Operations
```
Since this is all Provisioning logic I would expect the module (and file name) to be `provisioning` and not `operations`.

gmcculloug commented on app/models/manageiq/providers/redfish/physical_infra_manager/physical_server.rb:
```diff
@@ -1,7 +1,13 @@
       n_('Physical Server (Redfish)', 'Physical Servers (Redfish)', number)
     end
+
+    def kind
```
Same concern as https://github.com/ManageIQ/manageiq/pull/18573/files#r270238530.  The method name is too generic.

gtanzillo commented:
@gmcculloug Regarding
>@agrare @gtanzillo Lots of JS here and I am not sure where that belongs or who should review.

Adding @martinpovolny for further review. The toolbar and button overrides are leveraging the support we added last year for pluggable UI buttons and dialogs for provider operations so I believe this is fine. I'm not sure, however, how this would affect displaying the requests.

miha-plesko commented:
Hey guys, I've split the entire provisioning implementation in a series of PRs, all of them listed above. I would kindly ask for a review on all of them so we get it in eventually.

- @tadeboro I'd value your opinion from the Redfish perspective. As you can see I've split Redfish provisioning into three steps and almost each still has a TODO which I think we should handle later.
- @agrare I'd value your opinion on the internal state machine flow, especially the core one. We took VMware's internal state machine as an example. To be specific, from VMware we inherited approach that state machine's skeleton (in miq core) simply raises NotImplemented and requires provider to implement its own steps. I really like such approach because providers are then free to implement steps however they want. Anyway, looking forward to your opinion.
- @gmcculloug and @lfu I'd value your opinion on the Automate part ie. the external and internal state machine implementation. I think we touched common ground on the call so I'm confident it should be okay in general.

Looking forward to your comments!

tadeboro commented on app/models/manageiq/providers/redfish/physical_infra_manager/physical_server/provisioning.rb:
```diff
@@ -0,0 +1,42 @@
+
+    def mac_addresses(system)
+      system.EthernetInterfaces.Members.reduce(Set.new) do |acc, el|
+        acc.add(el.PermanentMACAddress).add(el.MACAddress)
```
A few days ago I learned the hard way that not all ethernet interfaces have both of those fields set. Maybe we can filter out the `nil`s with something similar to this:

    acc.merge([el.PermanentMACAddress, el.MACAddress].compact)

miha-plesko replied:
Thanks 👍 I've updated commit to properly handle even empty string values now, just in case some server thinks its a good joke to send them over.

agrare replied:
I think this method could be simplified a bit to be easier to read, here's what I'm thinking let me know if it matches how this currently works:

```ruby
system.EthernetInterfaces.Members.map { |iface| iface.MACAddress || iface.PermanentMACAddress }.compact
```

tadeboro replied:
> I've updated commit to properly handle even empty string values now, just in case some server thinks its a good joke to send them over.

@miha-plesko did you forgot to push the changes?

As for the `mac_addresses` method, its task is to collect all distinct MAC addresses from the Ethernet interfaces that are present on the system. For example, let us assume that we have four interfaces on our computer that report this back via Redfish API:

    eth0:
      MACAddress: mac1
      PermanentMACAddress: mac2
    eth1:
      MACAddress: mac3
    eth2:
      PermanentMACAddress: mac4
    eth3:
      MACAddress: mac5
      PermanentMACAddress: mac5

What we would like to get back is some iterable thing that contains all five MACs (mac1, mac2, ..., mac5). If we do naive `Members.flat_map { |iface| [iface.MACAddress, iface.PermanentMACAddress] }`, we will get back `["mac1", "mac2", "mac3", nil, nil, "mac4", "mac5", "mac5"]`. Now we need to filter out the `nil`s and remove duplicates. Calling `compact.to_set` on the `flat_map`'s return value would do the job, but I am not sure what would the most "rubyish" approach be.

agrare replied:
Oh okay I thought only one or the other could be set, sounds like `(Members.map(&:MACAddress) + Members.map(&:PermanentMACAddress)).reject(&:blank?)` is what you want then

miha-plesko replied:
@agrare we'd also need to remove the duplicates because as far as I know MAC and PermanenctMAC are usually the same (fyi: MAC is hardware MAC while PermanentMAC is optionally spoofed mac addr). So in the end we'd have

```ruby
def mac_addresses(system)
  [].tap do |macs|
    macs.concat(system.EthernetInterfaces.Members.map(&:MACAddress))
    macs.concat(system.EthernetInterfaces.Members.map(&:PermanentMACAddress))
    macs.reject(&:blank?).uniq  # get rid of nils, empty strings and duplicates
  end
end
```

Wdyt? I like both methods and can agree that this second one is more obvious.

@tadeboro the change I mentioned there is in L39 which removes all nils and emtpy strings from the set.

agrare replied:
@miha-plesko you can just tack a `.uniq` to the end of mine if you're worried about duplicate macs

gtanzillo replied:
Agree with @agrare suggestion, this is the more readable -
`system.EthernetInterfaces.Members.map { |iface| iface.MACAddress.presence || iface.PermanentMACAddress.presence }.compact.uniq`

tadeboro replied:
I really do not care what implementation we use, as long as this invariant holds true:

    require "redfish_client/resource"
    system = RedfishClient::Resource.new(nil, content: {
      "EthernetInterfaces" => {
        "Members" => [
          { "MACAddress" => "mac1", "PermanentMACAddress" => "mac2" },
          { "MACAddress" => "mac3"                                  },
          {                         "PermanentMACAddress" => "mac4" },
          { "MACAddress" => "mac5", "PermanentMACAddress" => "mac5" },
        ]
      }
    })
    mac_addresses(system).to_set == Set.new(%w[mac1 mac2 mac3 mac4 mac5])

miha-plesko commented:
@agrare I think we're good with this one as well now that the related core and automation PRs have been merged.

miha-plesko commented:
Been doing some cleanup and somehow deleted source branch of this PR as well. Have pushed the branch again, but PR can't be reassociated.. Therefore I've opened a new PR here https://github.com/ManageIQ/manageiq-providers-redfish/pull/62. I do apologize for the drama.
